### PR TITLE
Added French Canadian + Transifex hosting + i18n tooling refactor

### PR DESCRIPTION
**Pull request by OmarIthawi**
I needed French Canadian to be added but I noticed that this project is missing a couple of stuff:

 - Corrected weird links for locale
 - It's not been pushed to Transifex, so I've added it
 - Using `text/textjs` instead of `django/djangojs`
 - Not including `textjs.po` anymore with `textjs.po`
 - `underscore.po` was not being collected, now fixed and pushed it to Transifex within `textjs.po`
 - This :point_up: meant that the `generate_merge` step is not needed anymore
 - The pl language code isn't supported by Transifex, only `pl_PL`
 - Fixed Chinese, Polish and Japanese to work with Transifex
 - Arabic had dummy translations -- now removed
 - Added `make validate_translations` step to `tox.ini`
 - Deleted all po/mo/js language files and pulled new versions so no stale translations exists

----

#### Steps to Reproduce the Plurals Issue
 - Download this original [`textjs.po`](https://raw.githubusercontent.com/open-craft/problem-builder/484a8a7e8f1e6371db51891ba14a3958232d9c47/problem_builder/translations/zh_CN/LC_MESSAGES/textjs.po)
 - Go to https://www.transifex.com/open-edx/xblocks/problem-builder-js/
 - Click on Chinese (zh_CN)
 - Click on the `Upload file` button
 - Upload the `textjs.po` file you've just chose

#### Expected Result
To upload the file successfully

#### Actual Result
I'm getting the following error message:
<kbd></kbd>

The only workaround I know is to upload the [modified `textjs.po` version](https://raw.githubusercontent.com/open-craft/problem-builder/2e307ac8fee58877e2be78d710a36e801658e685/problem_builder/translations/zh_CN/LC_MESSAGES/textjs.po).

### CODE REVIEW

**Comment by OmarIthawi** [comment]
@pomegranited one more XBlock i18n PR :slightly_smiling_face:

File: problem_builder/translations/ar/LC_MESSAGES/text.po [hunk 1]
```diff
@@ -2,216 +2,211 @@
...
-#: problem_builder/answer.py problem_builder/instructor_tool.py
+#: answer.py:146 instructor_tool.py:140
 msgid "Long Answer"
-msgstr "مخرل شرسصثق"
+msgstr ""
 
-#: problem_builder/answer.py problem_builder/mixins.py
+#: answer.py:150 mixins.py:135
 msgid "Question ID (name)"
-msgstr "السؤال هي (رشوث)"
```
[the remark this answers is not in the document]
Comment (OmarIthawi) [reply problem_builder/translations/ar/LC_MESSAGES/text.po]: Thanks for asking @pomegranited. I've read the translations in Arabic and it appears to be incorrectly translated -- because I'm Arabic :man_shrugging: 

It's possible that another language like Urdu or Persian have been used under the code `ar`.

@xitij2000 @umarmughal824 what do you think?

**Comment by OmarIthawi** [comment]
Thanks @pomegranited for the review.

> Thanks for your PR @OmarIthawi -- I have some concerns:
> 
> > The pl language code isn't supported by Transifex, only pl_PL
> 
> That's fine -- but you need to run `make compile_translations` to ensure that the static JS files are generated for `pl_PL` (and your added `fr_CA`) too.

Will do that, thanks for the reminder.

> 
> > Chinese, Polish and Japanese have incorrect plural forms (according to Transifex)
> 
> Can you point me to where Transifex says this?

Sure, I've added the full issue details in the pull request description https://github.com/open-craft/problem-builder/pull/240#issue-301764527


We might need to unhardcode the `1` in the following file as well as similar ones to ensure proper number interpolation:

https://github.com/open-craft/problem-builder/blob/484a8a7e8f1e6371db51891ba14a3958232d9c47/problem_builder/templates/html/sb-review-score.html#L20-L22

What do you think?

**Comment by pomegranited** [comment]
@OmarIthawi 

> We might need to unhardcode the 1 in the following file as well as similar ones to ensure proper number interpolation:
> https://github.com/open-craft/problem-builder/blob/484a8a7e8f1e6371db51891ba14a3958232d9c47/problem_builder/templates/html/sb-review-score.html#L20-L22

No, I think the existing code is correct -- see docs for the `blocktrans` template tag: https://docs.djangoproject.com/en/1.11/topics/i18n/translation/#blocktrans-template-tag

I think the issue you're seeing with the invalid plurals is caused by [these lines in the Makefile](https://github.com/open-craft/problem-builder/blob/master/Makefile#L22-L23): 
```
sed -i'' -e 's/nplurals=INTEGER/nplurals=2/' $(EXTRACTED_TEXT)
sed -i'' -e 's/plural=EXPRESSION/plural=\(n != 1\)/' $(EXTRACTED_TEXT)
```
I don't think we can make this change across all the compiled translations, since each language will have its own set of plurals.  ref https://docs.transifex.com/formats/gettext#plural-forms-in-a-po-file

**Comment by OmarIthawi** [comment]
Thanks @pomegranited. Regarding the Chinese (and other langs) pluralization issue, I'm honestly not fully qualified to judge but I checked this [extensive article about pluralization](https://lingohub.com/blog/2019/02/pluralization/) suggests that there's only one form of plural forms in Chinese/Japanese.

> > We might need to unhardcode the 1 in the following file as well as similar ones to ensure proper number interpolation:
> > https://github.com/open-craft/problem-builder/blob/484a8a7e8f1e6371db51891ba14a3958232d9c47/problem_builder/templates/html/sb-review-score.html#L20-L22
> 
> No, I think the existing code is correct -- see docs for the `blocktrans` template tag: https://docs.djangoproject.com/en/1.11/topics/i18n/translation/#blocktrans-template-tag

I'm not fully qualified to judge, so I'll keep it as-is.

> I think the issue you're seeing with the invalid plurals is caused by [these lines in the Makefile](https://github.com/open-craft/problem-builder/blob/master/Makefile#L22-L23):
> 
> ```
> sed -i'' -e 's/nplurals=INTEGER/nplurals=2/' $(EXTRACTED_TEXT)
> sed -i'' -e 's/plural=EXPRESSION/plural=\(n != 1\)/' $(EXTRACTED_TEXT)
> ```

So, yeah that could be the issue and I think that we should remove those lines in favor of the more proper tool `$ i18n_tools extract/generate` like it's the case in other XBlocks.

However, the original [`textjs.po`](https://raw.githubusercontent.com/open-craft/problem-builder/484a8a7e8f1e6371db51891ba14a3958232d9c47/problem_builder/translations/zh_CN/LC_MESSAGES/textjs.po) file appears to be hand-made by a translator so it's likely a hand error than a script issue.

> I don't think we can make this change across all the compiled translations, since each language will have its own set of plurals. ref https://docs.transifex.com/formats/gettext#plural-forms-in-a-po-file

True, I'll switch to `i18n_tools` instead.

---

To summarize what I would like to do next:

 - Use `i18n_tools` when appropriate instead of `Makefile` commands.
 - Keep the hardcoded `1` in :point_down: since I don't know enough Japanese to judge 
   https://github.com/open-craft/problem-builder/blob/484a8a7e8f1e6371db51891ba14a3958232d9c47/problem_builder/templates/html/sb-review-score.html#L20-L22
 - Get the po files from Transifex and use whatever it gives us instead of the old ones

What do you think?

**Comment by pomegranited** [comment]
@OmarIthawi 

> Use i18n_tools when appropriate instead of Makefile commands.

The Makefile calls the right `i18n_tools` commands, so I would recommend just removing those `sed` lines, and carrying on with the make commands.

> Keep the hardcoded 1 in 👇 since I don't know enough Japanese to judge

👍 

> Get the po files from Transifex and use whatever it gives us instead of the old ones

👍 

And from [our previous discussion](https://github.com/open-craft/problem-builder/pull/240#issuecomment-515889472), ensure that you run `make compile_translations` over any updated translations.

Thank you !

**Comment by OmarIthawi** [comment]
I'm marking this as `(WIP)`. I made a small mess that I need to fix before unmarking as ready-for-review again.

**Comment by OmarIthawi** [comment]
@pomegranited I've cleaned up the PR and updated the description. Most notable, I've got back the `sed` stuff, `compile_translations` would break otherwise and it was very difficult to get `i18n_tool` understand the setup of this XBlock without doing too much refactor noise.

File: problem_builder/translations/ar/LC_MESSAGES/text.po (hunk 1 above)
[the remark this answers is not in the document]
Comment (xitij2000) [reply problem_builder/translations/ar/LC_MESSAGES/text.po]: Sorry for the delayed response! 

Unfortunately I don't have any knowledge of Arabic. My main involvement in previous PRs was in testing and ensuring that the translations showed up correctly. While I do often do a rough check on Google Translate for random passages, I don't really trust it since I know how easily it can mess up in my "native" language Hindi. 

Would you be able to help with a more accurate translation here? Would an inaccurate translation be better here than a missing one?

[the remark this answers is not in the document]
Comment (OmarIthawi) [reply problem_builder/translations/ar/LC_MESSAGES/text.po]: Thanks @xitij2000 for the review! Regarding the Arabic translation it's just pure random punches on the keyboard, that is incorrect. 

So I thought that there's no point of keeping it in a popular XBlock and confuse learners, instead English should be shown.

I would love to help with the Arabic, but I still have several issues to tackle in terms of XBlock i18n that is of higher priority.

However, I wouldn't be surprised to have it translated by the community on the next pull.

What do you think?

[the remark this answers is not in the document]
Comment (xitij2000) [reply problem_builder/translations/ar/LC_MESSAGES/text.po]: @omarithawi Sure. In that case I think it might be better to have a missing translation than an incorrect/misleading one I guess. 

The creators of the original PR for adding these translations themselves might be interested in fixing this. Will see if I can get in touch with them and get a response in time for this PR itself.

[the remark this answers is not in the document]
Comment (pomegranited) [reply problem_builder/translations/ar/LC_MESSAGES/text.po]: Thanks for contacting the original PR creators @xitij2000 . Any luck there?

[the remark this answers is not in the document]
Comment (xitij2000) [reply problem_builder/translations/ar/LC_MESSAGES/text.po]: @pomegranited Nope :-(

**Comment by OmarIthawi** [comment]
@pomegranited @xitij2000 since it's been difficult to follow up on the Arabic issue could you please suggest an alternative?

File: problem_builder/translations/ar/LC_MESSAGES/text.po (hunk 1 above)
[the remark this answers is not in the document]
Comment (pomegranited) [reply problem_builder/translations/ar/LC_MESSAGES/text.po]: No worries @xitij2000 .. Since @OmarIthawi has fixed up the translation pulling and compiling, it should be easy now for someone to go into Transifex and update the Arabic strings, and submit a new PR here.

[the remark this answers is not in the document]
Comment (xitij2000) [reply problem_builder/translations/ar/LC_MESSAGES/text.po]: Yup, that sounds fair enough!

File: problem_builder/translations/en/LC_MESSAGES/.gitignore [hunk 2]
```diff
@@ -0,0 +1,4 @@
+*
+!.gitignore
+!textjs.po
+!text.po
```
[the remark this answers is not in the document]
Comment (OmarIthawi) [reply problem_builder/translations/en/LC_MESSAGES/.gitignore]: Will do that, thanks @pomegranited!

**Comment by OmarIthawi** [comment]
> Yep, thanks for trying, @xitij2000 . I think we've waited long enough for a response, so can just proceed here, and if anyone else wants to change the Arabic strings, they can do so in Transifex and submit a follow-up PR.

:+1: 
 
> I can see you're already doing this @OmarIthawi , thank you! When I run `make pull_translations`, there are updated Arabic strings in `textjs.po`. Would you like to push those changes into this PR as well, or do you want to do this as a separate PR?

Done.

And yes, I translated few things, but the bulk of the strings still need to be translated. Soon another community member will translate it. Arabic is one of the active languages in the Open edX community.

> Also, thank you for taking the time to split the translations properly into `text` and `textjs`. That's working well.

:+1: 

> Have left one request for the `.gitignore` change, and could you also please update the [`VERSION`](https://github.com/appsembler/problem-builder/blob/i18n-bot/2019-07-27-164214/setup.py#L31) to `3.3.9`?

Done.

> Then I'll approve, merge, and create a new tag.

Thank you @pomegranited!

**Comment by OmarIthawi** [comment]
@pomegranited I think 39712fb is needed as well.

File: problem_builder/translations/ar/LC_MESSAGES/textjs.po [hunk 3]
```diff
@@ -1,52 +1,65 @@
...
+#: public/js/instructor_tool.js:429
 msgid "The report is currently being generated…"
-msgstr "فاث قثحخقف هس ذعققثرفمغ زثهرل لثرثقشفثي…"
-
-#: problem_builder/templates/html/mentoring_attempts.underscore
-msgid "You have used {num_used} of 1 submission."
-msgid_plural "You have used {num_used} of {max_attempts} submissions."
-msgstr[0] "استهلكت وظيفة الإرسال {num_used} من أصل 1."
-msgstr[1] "استهلكت وظيفة الإرسال {num_used} من أصل {max_attempts}."
+msgstr "جاري إنشاء التقرير..."
```
Comment (pomegranited) [inline problem_builder/translations/ar/LC_MESSAGES/textjs.po]: Thanks for pulling these update translations @OmarIthawi !  But you missed the compile step.  When I run `make compile_translations`, I get:
```
modified:   problem_builder/public/js/translations/ar/textjs.js
modified:   problem_builder/translations/ar/LC_MESSAGES/textjs.mo
```

File: Makefile [hunk 4]
```diff
@@ -13,6 +13,7 @@ upgrade:
 
 extract_translations: ## extract strings to be translated, outputting .po files
 	cd $(WORKING_DIR) && i18n_tool extract
+	mkdir -p $(EXTRACT_DIR)
```
[the remark this answers is not in the document]
Comment (OmarIthawi) [reply Makefile]: Sometimes I remove the directory for the purpose of cleaning it, so I'd like it to create the directory for me. It's a minor thing that I will remove.

File: problem_builder/translations/ar/LC_MESSAGES/textjs.po (hunk 3 above)
Comment (OmarIthawi) [reply problem_builder/translations/ar/LC_MESSAGES/textjs.po]: That's always easy to forget. Thanks for the double check @pomegranited.

**Comment by OmarIthawi** [comment]
Thank you again @pomegranited, I've squashed the commits. It now should be ready to merge.

**Comment by pomegranited** [comment]
@OmarIthawi  merged, and new tag created: [v3.3.9](https://github.com/open-craft/problem-builder/tree/v3.3.9)

**Comment by OmarIthawi** [comment]
Thank you @pomegranited!
